File: python/lista03.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf
from common import TABLES, wooldridge_data, savefig

# Q4 - ceosal1
try:
    ceo = wooldridge_data('ceosal1')
    mod = smf.ols('np.log(salary) ~ np.log(sales) + roe + ros', data=ceo).fit()
    (TABLES / 'lista03_q04_ceosal1.txt').write_text(mod.summary().as_text(), encoding='utf-8')

    ros_grid = np.linspace(ceo['ros'].min(), ceo['ros'].max(), 100)
    base = ceo[['sales', 'roe']].mean()
    y_grid = (mod.params['Intercept'] + mod.params['np.log(sales)']*np.log(base['sales'])
              + mod.params['roe']*base['roe'] + mod.params['ros']*ros_grid)
    plt.figure(figsize=(7, 4))
    plt.scatter(ceo['ros'], np.log(ceo['salary']))
    plt.plot(ros_grid, y_grid)
    plt.xlabel('Retorno sobre vendas, ROS')
    plt.ylabel('log(salário)')
    plt.title('Lista 3 - Questão 4: ROS e salário dos CEOs')
    savefig(plt, 'lista03_q04_ceosal1_ros_lsalary')
except Exception as e:
    (TABLES / 'lista03_q04_ceosal1_erro.txt').write_text(str(e), encoding='utf-8')

# Q6 - return
try:
    ret = wooldridge_data('return').rename(columns={'return':'ret'})
    m1 = smf.ols('ret ~ dkr + eps + salary + netinc', data=ret).fit()
    (TABLES / 'lista03_q06_return_nivel.txt').write_text(m1.summary().as_text(), encoding='utf-8')
    ret = ret.assign(lnetinc=np.log(ret['netinc']), lsalary=np.log(ret['salary']))
    m2 = smf.ols('ret ~ dkr + eps + lsalary + lnetinc', data=ret).fit()
    (TABLES / 'lista03_q06_return_log.txt').write_text(m2.summary().as_text(), encoding='utf-8')

    plt.figure(figsize=(7, 4))
    plt.scatter(ret['ret'], m2.fittedvalues)
    lims = [min(ret['ret'].min(), m2.fittedvalues.min()), max(ret['ret'].max(), m2.fittedvalues.max())]
    plt.plot(lims, lims)
    plt.xlabel('Retorno observado')
    plt.ylabel('Retorno ajustado')
    plt.title('Lista 3 - Questão 6: observado versus ajustado')
    savefig(plt, 'lista03_q06_return_observado_ajustado')
except Exception as e:
    (TABLES / 'lista03_q06_return_erro.txt').write_text(str(e), encoding='utf-8')

print('Lista 3 concluída.')

File: R/lista03.R
# Carrega o setup de forma robusta, com ou sem run_all.R
if (is.null(getOption("econ_r_dir")) || is.null(getOption("econ_root"))) {
  .locate_r_dir_local <- function() {
    of <- tryCatch(sys.frame(1)$ofile, error = function(e) NA_character_)
    if (!is.null(of) && length(of) == 1 && is.character(of) && !is.na(of) && nzchar(of)) {
      return(dirname(normalizePath(of, winslash = "/", mustWork = FALSE)))
    }
    cwd <- normalizePath(getwd(), winslash = "/", mustWork = TRUE)
    if (basename(cwd) == "R" && file.exists(file.path(cwd, "00_setup.R"))) return(cwd)
    if (file.exists(file.path(cwd, "R", "00_setup.R"))) return(file.path(cwd, "R"))
    stop("Não foi possível localizar a pasta R do pacote.")
  }
  source(file.path(.locate_r_dir_local(), "00_setup.R"), chdir = TRUE)
} else {
  source(file.path(getOption("econ_r_dir"), "00_setup.R"), chdir = TRUE)
}

# Q4 ceosal1
data('ceosal1', package='wooldridge')
mod_ceo <- lm(log(salary) ~ log(sales) + roe + ros, data=ceosal1)
capture.output(summary(mod_ceo), file=file.path(TABLES, 'R_lista03_q04_ceosal1.txt'))
save_plot('R_lista03_q04_ceosal1_ros_lsalary', {
  plot(ceosal1$ros, log(ceosal1$salary), xlab='Retorno sobre vendas, ROS', ylab='log(salário)', main='ROS e salário dos CEOs')
  ros_grid <- seq(min(ceosal1$ros, na.rm=TRUE), max(ceosal1$ros, na.rm=TRUE), length.out=100)
  base_sales <- mean(ceosal1$sales, na.rm=TRUE); base_roe <- mean(ceosal1$roe, na.rm=TRUE)
  y_grid <- coef(mod_ceo)[1] + coef(mod_ceo)[2]*log(base_sales) + coef(mod_ceo)[3]*base_roe + coef(mod_ceo)[4]*ros_grid
  lines(ros_grid, y_grid)
})

# Q6 return
tryCatch({
  data(list='return', package='wooldridge')
  ret <- get('return')
  names(ret)[names(ret) == 'return'] <- 'ret'
  mod_ret1 <- lm(ret ~ dkr + eps + salary + netinc, data=ret)
  capture.output(summary(mod_ret1), file=file.path(TABLES, 'R_lista03_q06_return_nivel.txt'))
  ret$lnetinc <- log(ret$netinc)
  ret$lsalary <- log(ret$salary)
  mod_ret2 <- lm(ret ~ dkr + eps + lsalary + lnetinc, data=ret)
  capture.output(summary(mod_ret2), file=file.path(TABLES, 'R_lista03_q06_return_log.txt'))
  save_plot('R_lista03_q06_return_observado_ajustado', {
    plot(ret$ret, fitted(mod_ret2), xlab='Retorno observado', ylab='Retorno ajustado', main='Retorno observado versus ajustado')
    abline(0, 1)
  })
}, error=function(e) {
  writeLines(as.character(e), file.path(TABLES, 'R_lista03_q06_return_erro.txt'))
})
cat('Lista 3 R concluída.\n')